### PR TITLE
Add README and scripts for kubernetes deployment

### PR DESCRIPTION
**Pull request by samanehsan**
- Add notes on how to create a new kubernetes listener deployment and how to update an existing deployment
- Add scripts to populate listener config secret and tls secret

### CODE REVIEW

File: kubernetes/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+##Create a new kubernetes listener deployment
```
Comment (rexwangcc) [inline kubernetes/README.md]: I must say, this is super useful!

File: kubernetes/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+16. Add new record set in AWS
+
+##Update existing deployment 
+
+To deploy new code to an existing deployment environment: 
+1. Build docker image: `bash build_docker.sh dev test`
+2. Push docker image to gcloud: `gcloud docker -- push <image_url>`
+3. Update `listener-deployment.yaml` image with the pushed docker image
+4. Set listener config secret `bash populate-listener-config-secret.sh`
+5. Update deployment: `kubectl apply -f listener-deployment.yaml --record`
```
Comment (rexwangcc) [inline kubernetes/README.md]: I must say, this is super useful!!

Comment (ambrosejcarr) [reply kubernetes/README.md]: Ahhh this is great. 👍

File: kubernetes/README.md (hunk 1 above)
Comment (dshiga) [inline kubernetes/README.md]: I think you need a space after the #'s for GitHub to display it correctly as a header.

File: kubernetes/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+##Create a new kubernetes listener deployment
+
+###Requirements:
+- Install [gcloud](https://cloud.google.com/sdk/)
+- Install [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/): `gcloud components install kubectl`
+
+###Setup:
+1. Create gcloud project for the new deployment (Each deployment corresponds to a separate gcloud project. e.g. the staging deployment uses `broad-dsde-mint-staging`)
+2. Create listener cluster `gcloud container clusters create listener`
+3. Add `config.json` and `bucket-reader-key.json` to /etc/secondary-analysis/
```
Comment (dshiga) [inline kubernetes/README.md]: Might be helpful to clarify that this is on your local machine - user will need to make the dir /etc/secondary-analysis and put the files there.

File: kubernetes/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+4. Add the wdl files listed in `config.json` to a bucket in the new project and update the wdl paths
```
Comment (dshiga) [inline kubernetes/README.md]: Might want to clarify a bit like:
Upload wdl, zip and json files needed by the listener to a bucket in the new project and edit config.json to refer to those paths.

File: kubernetes/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+5. Set listener config secret `bash populate-listener-config-secret.sh`
+6. Add `fullchain.pem` and `privkey.pem` to /etc/secondary-analysis/
+7. Set tls secret: `bash populate-tls-secret.sh`
+8. Build docker image: `bash build_docker.sh dev test`
```
Comment (dshiga) [inline kubernetes/README.md]: the command should be bash build_docker.sh [env] [tag] where env is the environment to deploy to and tag is a tag for the docker image.

File: kubernetes/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+9. Push docker image to gcloud: `gcloud docker -- push <image_url>`
+10. Update `listener-deployment.yaml` image with the pushed docker image
+11. Create deployment: `kubectl create -f listener-deployment.yaml --record --save-config`
+12. Create service: `kubectl create -f listener-service.yaml --record --save-config`
+13. Create ingress: `kubectl create -f listener-ingress.yaml --record --save-config`
+14. Create static ip address: `gcloud compute addresses create listener --global`
+15. In the Google Cloud Platform console, go to the Network Services/Cloud DNS page and add a new record set for the static ip address
```
Comment (dshiga) [inline kubernetes/README.md]: more specifically, add an "A" record named listener101.mint-[env].broadinstitute.org that points to the static ip address.

File: kubernetes/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+16. Add new record set in AWS
```
Comment (dshiga) [inline kubernetes/README.md]: In AWS, go to route53, click on hosted zones, then data.humancellatlas.org. Add or modify a CNAME record named pipelines.[env].data.humancellatlas.org and set the value to listener101.mint-[env].broadinstitute.org. For prod this should just be pipelines.data.humancellatlas.org.

File: kubernetes/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+##Update existing deployment 
+
+To deploy new code to an existing deployment environment: 
+1. Build docker image: `bash build_docker.sh dev test`
```
Comment (dshiga) [inline kubernetes/README.md]: should be bash build_docker.sh [env] [tag]

File: kubernetes/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+11. Create deployment: `kubectl create -f listener-deployment.yaml --record --save-config`
+12. Create service: `kubectl create -f listener-service.yaml --record --save-config`
+13. Create ingress: `kubectl create -f listener-ingress.yaml --record --save-config`
+14. Create static ip address: `gcloud compute addresses create listener --global`
+15. In the Google Cloud Platform console, go to the Network Services/Cloud DNS page and add a new record set for the static ip address
+16. Add new record set in AWS
+
+##Update existing deployment 
+
+To deploy new code to an existing deployment environment:
```
Comment (dshiga) [inline kubernetes/README.md]: Should add some steps at the beginning to switch to the correct environment:
gcloud container clusters --project [env] get-credentials listener
kubectl config get-contexts, then find the context name for the environment you want
kubectl config set-context [name] where [name] is the context name, e.g. gke_broad-dsde-mint-dev_us-central1-b_listener.

**Review by dshiga — APPROVED** [review]
Looks great! I commented on individual lines of the README with suggestions.

**Comment by samanehsan** [comment]
Thanks for your suggestions @dshiga! I updated the README with those changes

**Comment by kbergin** [comment]
Total side note question @dshiga @samanehsan - is this README intended to be so others could stand up their own green box if they wanted to? If so, is it possible to make it simpler for them to do? Maybe with some script that does all the steps outlined for them? I'm assuming if the answer to that is yes, that is considered part of our future automated deployment efforts. Total curiosity here/thinking about the backlog, not expecting it to be updated to do this now.

**Comment by dshiga** [comment]
This was mainly meant to document our current process for deploying for our own reference. Deploying will definitely get easier when we overhaul things and create Jenkins jobs. There might not ever be a single script that does it all in one fell swoop, but rather a collection of scripts and possibly some manual steps. App engine would be one way to deploy more easily but comes with too many limitations for us to want to do it that way.

**Comment by kbergin** [comment]
@dshiga thanks for explaining!

**Pull request merged**